src/main/java/com/google/devtools/build/lib/analysis/CommonPrerequisiteValidator.java

```diff
@@ -105,21 +108,46 @@ private void validateDirectPrerequisiteVisibility(
     if (!toolCheckAtDefinition
         || !attribute.isImplicit()
         || attribute.getName().equals(RuleClass.CONFIG_SETTING_DEPS_ATTRIBUTE)
-        || rule.getRuleClassObject().getRuleDefinitionEnvironmentLabel() == null) {
+        || !forStarlarkRuleOrAspect(context)) {
       // Default check: The attribute must be visible from the target.
       if (!context.isVisible(prerequisite.getConfiguredTarget())) {
         handleVisibilityConflict(context, prerequisite, rule.getLabel());
       }
     } else {
-      // For implicit attributes, check if the prerequisite is visible from the location of the
-      // rule definition
-      Label implicitDefinition = rule.getRuleClassObject().getRuleDefinitionEnvironmentLabel();
+      // For implicit attributes of Starlark rules or aspects, check if the prerequisite is visible
+      // from the location of the definition that declares the attribute. Only perform this check
+      // for the current aspect.
+      Label implicitDefinition;
+      Aspect mainAspect = context.getMainAspect();
+      if (mainAspect != null) {
+        if (!mainAspect.getDefinition().getAttributes().containsKey(attrName)) {
+          // Only verify visibility of implicit dependencies of the current aspect. Implicit
+          // dependencies of other aspects as well as the rule itself are checked when they are
+          // evaluated.
+          return;
+        }
+        StarlarkAspectClass aspectClass = (StarlarkAspectClass) mainAspect.getAspectClass();
+        implicitDefinition = aspectClass.getExtensionLabel();
+      } else {
+        // Never null since we already checked that the rule is a Starlark rule.
+        implicitDefinition =
+            checkNotNull(rule.getRuleClassObject().getRuleDefinitionEnvironmentLabel());
+      }
       if (!RuleContext.isVisible(implicitDefinition, prerequisite.getConfiguredTarget())) {
         handleVisibilityConflict(context, prerequisite, implicitDefinition);
       }
     }
   }
 
+  private static boolean forStarlarkRuleOrAspect(RuleContext.Builder context) {
```

Comment:
Readability: move `forStarlarkRuleOrAspect` into the RuleContext.Builder.
Rename `forStarlarkRuleOrAspect` to `isStarlarkRuleOrAspect` (I see there is preexsiting use of `for`, but I think is would make more sense)